### DISCUSSION TITLE
Low Poly Terrains

### DISCUSSION THREAD

**Discussion by TokisanGames** [TokisanGames/Terrain3D#435] [Extras]
![image](https://github.com/user-attachments/assets/67ed9551-52bd-4bde-afe4-426512bb3f3d)

To make a low poly terrain you need:
* Large polygons
* Flat shaded normals for each polygon, not interpolated normals or smooth shading
* Simple textures or colors

Instructions:
- Set mesh_vertex_density to 10+
- Set mesh_size to 64 (See below)
- Sculpt
- Enable the override shader and use minimum.gdshader

This is the updated shader snippet to calculate screen space normals:
```glsl
    vec3 s_normal = normalize(cross(dFdy(VERTEX), dFdx(VERTEX)));
    vec3 w_normal = mat3(INV_VIEW_MATRIX) * s_normal;
    vec3 w_tangent = cross(w_normal, vec3(0, 0, 1));
    vec3 w_binormal = cross(w_normal, w_tangent);
    NORMAL = s_normal;
    TANGENT = mat3(VIEW_MATRIX) * w_tangent;
    BINORMAL = mat3(VIEW_MATRIX) * w_binormal;
```

Customize the shader as you see fit. You can use the default shader for painting, but it is overkill and has features that should be stripped out.

This method currently has an artifact on the LOD boundary which might be addressed by future improvements. Expanding mesh_size pushes it out further.

See #422 for a more complete example, but use the updated snippet here.

See #434 for gathered normals, instead of screenspace normals shown here. From @Xtarsia: "I think the screenspace method is better for [this] case."

**Comment by esklarski** [comment]
For the record all I had to do was replace the snippet in the default shader and it worked nicely:
![Screenshot from 2024-09-19 12-56-52](https://github.com/user-attachments/assets/7b48e299-b8f8-47d6-91fe-da0f4d8471d5)
This is with 0.9.2

I'm curious if we know where the root of the artifacting at mesh edges is? Is it that when there is no adjoining vertex (ie the edge verticies) the values calculated above get wonky?

**Comment by TokisanGames** [comment]
Low poly shaders can now be found in the /extras directory. Added in #609.

The artifact between lods will be addressed by circular lods and/or geomorphing per @Xtarsia.
